src/RMPI/pointtopoint.jl

```diff
@@ -1,64 +1,111 @@
 """
-function mpi_default(data, comm = mpi_comm(), root = mpi_root)
+function mpi_point_to_point(data, comm = mpi_comm(), root = mpi_root)
     MPI.Initialized() || error("MPI needs to be initialised first.")
     np = MPI.Comm_size(comm)
     id = MPI.Comm_rank(comm)
-    s = MPIDefault(np, id, comm)
+    s = MPIPointToPoint(pairtype(data), np, id, comm)
     return MPIData(data, comm, root, s)
 end
 
 """
-    MPIDefault(nprocs, my_rank, comm)
-Strategy for point-to-point MPI communication.
+    MPIPointToPoint{N,A}
+
+Point-to-point communication strategy. Uses circular communication using `MPI.Send` and
+`MPI.Recv!`.
+
+# Constructor
+
+* `MPIPointToPoint(::Type{P}, np, id, comm)`: Construct an instance with pair type `P` on
+  `np` processes with current rank `id`.
+
 """
-struct MPIDefault <: DistributeStrategy
+struct MPIPointToPoint{P,N} <: DistributeStrategy
     np::Int32
     id::Int32
     comm::MPI.Comm
+    datatype::MPI.Datatype
+    buffers::NTuple{N,Vector{P}}
+end
+function MPIPointToPoint(::Type{Pair{K,V}}, np, id, comm) where {K,V}
+    P = Pair{K,V}
+    datatype = MPI.Datatype(P)
+    return MPIPointToPoint{P,np}(np, id, comm, datatype, ntuple(_ -> P[], np))
+end
+"""
+    recvbuff(s::MPIPointToPoint)
+
+Get the receive buffer.
+"""
+function recvbuff(s::MPIPointToPoint)
+    return s.buffers[s.id + 1]
 end
+"""
+    sendbuff(s::MPIPointToPoint, id)
 
-function Rimu.sort_into_targets!(target, bufs::Vector{Vector{P}}, lens, ::Type{P}, s::MPIDefault) where P
-    # use standard MPI message passing communication
-    # use ring structure for sending around data with blocking communications:
-    # first receive from lower ranks, then send, then recieve from higher ranks
-    # receiving from lower ranks
-    for r = 0:(s.id-1)
-        status = MPI.Probe(r, 0, s.comm)
-        count = MPI.Get_count(status, P) # how many pairs are ready
-        rbuf = Vector{P}(undef,count) # allocate buffer of correct size
-        MPI.Recv!(rbuf, r, 0, s.comm)
-        for (key, val) in rbuf # sort into target dict right away
-            target[key] += val
-        end
-    end
-    # send all buffer entries moving to higher ranks
-    for tr in (s.id + 1) : (s.np - 1)
-        # s.id == 0 && println("sending to rank $tr")
-        sstat = MPI.Send(view(bufs[tr],1:lens[tr]), tr, 0, s.comm)
-        # println("sent from $(s.id) to $tr with status $sstat")
+Get the send buffer associated with `id`.
+"""
+function sendbuff(s::MPIPointToPoint, id)
+    return s.buffers[id + 1]
+end
+"""
+    receive!(target, s::MPIPointToPoint, id)
+
+Recieve from rank with `id` and move recieved values to `target`.
+"""
+function receive!(target, s::MPIPointToPoint{P}, id) where P
+    status = MPI.Probe(id, 0, s.comm)
+    count = MPI.Get_count(status, s.datatype)
+    resize!(recvbuff(s), count)
+    rb = recvbuff(s)
+    MPI.Recv!(MPI.Buffer(rb, length(rb), s.datatype), id, 0, s.comm)
```

Comment:
I see that you are reusing the same buffers for receiving and for sending. Very clever solution!